SCSS : mixins pour les media-queries adaptatives

AmauryCarrade opened the pull request:
Cette PR ajoute plusieurs _mixins_ permettant d'inclure les _media-queries_ servant à adapter le site à divers appareils. Elle met également à jour les points d'arrêt des différentes tailles d'écran, ainsi que leurs noms qui n'étaient, à mon humble avis, pas très pertinents.

Elle s'inscrit dans un objectif d'amélioration globale du code source de l'interface de Zeste de Savoir.

### Pourquoi ce changement ?

L'objectif est double : **clarifier le code** et **simplifier le développement**.

On passe de ceci : 

```scss
@media screen and #{$media-mobile-tablet} {
  // ...
}
```

à ceci : 

```scss
@include until-desktop {
  // ...
}
```

Le code est beaucoup plus clair : on voit immédiatement sans avoir à scanner la _media-query_ que l'on a affaire à un bloc concernant les appareils mobiles et tablettes (plus petits qu'un ordinateur de bureau). Il est également plus simple à écrire, car il n'est plus nécessaire de taper toute la _media-query_ et de se souvenir des noms des variables, à entourer de `#{}` pour que ce soit interprété littéralement par SCSS… non, un `@include mobile` et basta.

Il est également plus représentatif de la réalité, car j'ai renommé les points d'arrêt : 
- `mobile-tablet` devient `touch`, afin d'expliciter le fait que cette règle va concrètement viser les appareils tactiles, pour lesquels la réflexion d'interface n'est pas forcément la même que pour un appareil utilisé à l'aide d'une souris ;
- la série des `wide`, `extra-wide`, `mega-wide` est remplacée par des noms plus explicites et moins _dramatiques_ : `desktop`, `wide` et `fullhd` ;
- vu que les _mixins_ inutilisées disparaissent totalement une fois compilées, j'en ai ajouté quelques unes de plus afin de simplifier d'avance les futurs développements. Chaque intervalle est plus ou moins sélectionnable : pour chaque type d'écran, le type uniquement (par exemple `tablet-only`), le type plus toutes les tailles supérieures (par exemple `desktop`), et le type plus toutes les tailles inférieures (par exemple `until-wide` ou le cas particulier `touch`, qui serait équivalent à `until-desktop`).

Enfin dans l'objectif de simplification, **tout ceci est désormais documenté** — ce n'était pas le cas avant.

Les valeurs des points d'arrêt ont été légèrement modifiées (chacune gagne une centaine de pixels, selon les standards les plus utilisés aujourd'hui), afin de suivre l'évolution des différentes tailles d'écran. Ce point peut être trivialement annulé, afin de profiter du confort des _mixins_ avec les anciens points d'arrêt.

_Les styles d'erreur (`errors/scss/*.scss`) ne sont pas concernées par ce changement, étant isolées du reste des feuilles de style. J'ai également laissé deux media-queries aux valeurs très spécifiques dans `layout/_header`, car elles ne rentraient de toute façon pas dans un point d'arrêt._

### Les points d'arrêt et _mixins_ ajoutées

_(Extrait de la documentation écrite pour l'occasion.)_

#### Points d'arrêt

- ``$tablet`` : largeur minimale des tablettes (769 pixels, auparavant 760).
- ``$desktop`` : largeur minimale d'un écran d'ordinateur classique (1024 pixels, auparavant 960).
- ``$wide`` : largeur minimale d'un écran large (1216 pixels, auparavant 1140).
- ``$fullhd`` : largeur minimale d'un écran full-HD (1408 pixels, auparavant 1360).

#### _Mixins_

- ``mobile`` : les appareils mobiles.
- ``tablet`` : les tablettes, et toutes les tailles au dessus.
- ``tablet-only`` : les tablettes, uniquement.
- ``touch`` : les mobiles et les tablettes.
- ``desktop`` : les ordinateurs, toutes tailles d'écran.
- ``desktop-only`` : les ordinateurs, mais pas ceux de grande taille.
- ``until-wide`` : les mobiles, tablettes, et ordinateurs pas trop grands.
- ``wide`` : les ordinateurs, à partir de ceux ayant un grand écran.
- ``wide-only`` : les ordinateurs ayant un grand écran, mais pas un écran full-HD.
- ``until-fullhd`` : tous les appareils, sauf ceux ayant un écran full-HD.
- ``fullhd`` : les ordinateurs ayant un écran full-HD. Les téléviseurs, aussi.

### Contrôle qualité

Vérifier que le site s'affiche toujours correctement dans toutes les tailles d'écran.

L'usage de _mixins_ ne change pas le code CSS généré ; par contre, la modification des points d'arrêt pourrait avoir des effets secondaires imprévus, même si j'ai regardé un peu partout et que je n'ai rien vu de particulier.

AmauryCarrade committed:
Mixins pour les media-queries adaptatives

Ajout de quelques mixins afin de simplifier l'inclusion
de media-queries utilisées pour adapter le site aux
différentes tailles d'écran. Il n'est plus nécessaire
d'écrire les @media en entier ; un @import mobile {}
suffit (ou autres, bien sûr).

Au passage, les noms des différents points d'arrêt ont
été améliorés, et leurs valeurs, ajustées à la réalité
du marché actuel.

Situphen reviewed (commented):
Super proposition ! J'ai juste quelques points sur lesquels pinailler.

J'aimerais qu'on renomme `fullhd` en autre chose (par exemple `extra-wide`) car cette appelation est utilisée spécifiquement pour du 1 920 × 1 080 pixels alors que dans ta proposition il est question de 1408 px.

Je n'aime pas trop `touch` car pour moi ça fait référence au tactile hors ici il est question de la définition de l'écran (n'oublions pas les ordinateurs tactiles). On pourrait remettre le nom d'origine `mobile-tablet` qui est beaucoup plus clair. Ou alors on peut suivre la logique `X`, `X-only` et `until-X` et le renommer `until-desktop` (auxquel cas renommer aussi `mobile` en `until-tablet`).

AmauryCarrade commented:
Merci pour ton retour !

> J'aimerais qu'on renomme `fullhd` en autre chose (par exemple `extra-wide`) car cette appelation est utilisée spécifiquement pour du 1 920 × 1 080 pixels alors que dans ta proposition il est question de 1408 px.

J'approuve. `extra-wide` me va.

> Je n'aime pas trop `touch` car pour moi ça fait référence au tactile hors ici il est question de la définition de l'écran (n'oublions pas les ordinateurs tactiles). On pourrait remettre le nom d'origine `mobile-tablet` qui est beaucoup plus clair. Ou alors on peut suivre la logique `X`, `X-only` et `until-X` et le renommer `until-desktop` (auxquel cas renommer aussi `mobile` en `until-tablet`).

Ça se défend, pour `touch`. Je veux bien te l'accorder. Par contre vu que `mobile` sera très utilisé, je trouve que faire une exception dans la logique (plutôt que `until-tablet`) fait sens, et c'est à mon sens plus intuitif. Quelque chose comme `mobile` → `tablet` → `until-desktop` me semble mieux.

AmauryCarrade committed:
Mixins responsives : fullhd → extra-wide ; touch → until-desktop.

Mise à jour de la documentation, et ajouts d'informations et d'indications plus
explicites à propos des largeurs des différentes mixins.

AmauryCarrade commented:
J'ai appliqué les propositions et actualisé la documentation. Il ne reste plus qu'à vérifier que tout est bon (je n'ai pas vu de problème, mais l'auteur d'une PR ne va pas faire la QA lui-même…), et ce sera fusionnable.

AmauryCarrade committed:
GitHub check is stuck even if Travis is green

philippemilink commented:
Rapport de QA
---------------

Je viens de tester dans la vue adaptative de Firefox, je n'ai pas remarqué de problème.

**QA OK** ✔️ 

(mais je ne merge pas, peut-être que quelqu'un veut faire la QA sur un vrai périphérique avec un petit écran)

AmauryCarrade committed:
Merge branch 'dev' into feature-responsive-mixins

Situphen reviewed (approved):
**QA OK** pour moi aussi !
